File: php/phplog/src/NumberDemo.php
<?php

namespace Laixhe\Phplog;

/**
 * 数值类型示例：格式化输出、整数溢出行为、类型转换。
 * 对应 Rust rustlog/src/number.rs 与 Go golog 的数值示例。
 *
 * 前置知识：
 * - PHP 整数是 64 位有符号（PHP_INT_MAX = 9223372036854775807）
 * - PHP 整数溢出时不会抛异常，而是「静默提升为 float」（丢失精度），
 *   这是与 Java Math.addExact 最大的不同，业务上需要自行检查
 * - 需要任意精度时使用 BCMath / GMP 扩展
 */
final class NumberDemo
{
    // ============ 数值转字符串与格式化输出 ============
    public static function numberToString(): void
    {
        $i = 666;
        $f1 = 88.888;
        $f2 = 88.0;

        // 基础转字符串（对应 i.to_string()）
        echo "i={$i}", PHP_EOL; // 结果：i=666

        // 精度控制（四舍五入）
        printf("f1=%.2f" . PHP_EOL, $f1); // 结果 f1=88.89
        printf("f2=%.2f" . PHP_EOL, $f2); // 结果 f2=88.00

        // ===== 更多格式化方式（新手学习重点）=====

        // 十六进制 / 八进制 / 二进制
        printf("666 hex=0x%X  octal=0o%o  binary=0b%s" . PHP_EOL, $i, $i, decbin($i));
        // 结果：666 hex=0x29A  octal=0o1232  binary=0b1010011010

        // 前导零填充 + 宽度控制：%08d 表示「右对齐，总宽度 8，不足补 0」
        printf("666 with leading zeros: %08d" . PHP_EOL, $i);
        // 结果：666 with leading zeros: 00000666

        // 对齐：%-10d 左对齐，%10d 右对齐（默认）
        printf("left=|%-10d| right=|%10d|" . PHP_EOL, $i, $i);
        // 结果：left=|666       | right=|       666|

        // 正负号显式显示
        printf("positive= %+d  negative= %+d" . PHP_EOL, 666, -888);
        // 结果：positive= +666  negative= -888

        // 千分位分组
        echo 'grouping: ', number_format(1_234_567), PHP_EOL; // 1,234,567
    }

    // ============ 整数溢出行为（与 Java/Rust 对比）============
    // PHP 与 Go 类似：整数溢出「静默回绕/提升」，需要自己检查边界
    public static function overflow(): void
    {
        $x = PHP_INT_MAX;

        // 1) PHP 整数溢出：提升为 float，不抛异常（区别于 Java Math.addExact）
        $r = $x + 1;
        echo 'PHP_INT_MAX + 1 = ', var_export($r, true), '（溢出提升为 float）', PHP_EOL;

        // 2) 检查是否溢出：先把结果与最大值比较
        //    推荐：计算前判断「加上这个数是否会超过上限」
        $add = 1;
        if ($add > PHP_INT_MAX - $x) {
            echo "addExact 模拟: MAX + 1 = 溢出了（超出 int 范围）", PHP_EOL;
        } else {
            echo 'addExact 模拟: MAX + 1 = ', $x + $add, PHP_EOL;
        }

        // 3) saturating：饱和运算，溢出时取类型最大值（对应 Rust saturating_add）
        $sat = min($x + 1, PHP_INT_MAX);
        echo 'saturating_add: MAX + 1 = ', $sat, '（饱和，卡在 PHP_INT_MAX）', PHP_EOL;

        // 4) 小整数溢出：8 位无符号 255 + 1 = 256（PHP 无 u8，直接正常计算）
        $b = 255;
        echo '255 + 1 = ', $b + 1, '（PHP 整数默认 64 位，不回绕）', PHP_EOL;

        // 5) 除零会抛 DivisionByZeroError（PHP 8+，区别于早期版本的 INF/NAN）
        try {
            $z = 1 / 0;
        } catch (\DivisionByZeroError $e) {
            echo '1 / 0 = 抛 DivisionByZeroError（区别于 IEEE754 的 INF）', PHP_EOL;
        }
        // PHP 8 起浮点除以 0.0 同样抛异常（不再是 INF）
        try {
            $f = 1 / 0.0;
        } catch (\DivisionByZeroError $e) {
            echo '1 / 0.0 = 同样抛 DivisionByZeroError（PHP 8 起任何除零都抛异常）', PHP_EOL;
        }
    }

    // ============ 数值类型转换 ============
    // PHP 是弱类型语言，类型转换非常宽松，需要理解隐式转换规则
    public static function typeConversion(): void
    {
        // --- 1. 隐式转换（宽松）：int + string 自动转数值 ---
        $sum = 10 + '5';
        var_dump($sum); // int(15)

        // --- 2. 强制转换（cast）---
        $small = 10;
        $big = (float) $small; // int -> float
        var_dump($big);        // float(10.0)

        $pi = 3.99;
        $truncated = (int) $pi; // 浮点转整数：向零截断
        var_dump($truncated);   // int(3)

        $rounded = (int) round($pi); // 四舍五入后再转
        var_dump($rounded);          // int(4)

        // 字符串转数值
        $i = (int) '666';
        var_dump($i); // int(666)

        // --- 3. 类型判断与严格比较 ---
        var_dump(is_int(666));    // true
        var_dump(is_float(88.8)); // true
        // PHP 8 之前 == 是宽松比较，PHP 8 之后建议使用 ===
        var_dump(1 == '1');  // true（宽松比较，自动转换）
        var_dump(1 === '1'); // false（严格比较，类型也相同才相等）

        // --- 4. 字符串解析（对应 Rust string_parse）---
        $i2 = intval('666');
        echo "intval('666') = {$i2}", PHP_EOL; // 666
        $f2 = floatval('88.88');
        echo "floatval('88.88') = {$f2}", PHP_EOL; // 88.88
        // 解析失败：PHP 不会抛异常，而是返回 0（注意！）
        $bad = intval('not_a_number');
        echo "intval('not_a_number') = {$bad}（解析失败返回 0，不抛异常）", PHP_EOL;
        // 需要严格校验时用 filter_var
        $checked = filter_var('666', FILTER_VALIDATE_INT);
        var_dump($checked); // int(666)
        var_dump(filter_var('not_a_number', FILTER_VALIDATE_INT)); // false
    }
}

File: java/javalog/src/main/java/com/laixhe/javalog/demo/NumberDemo.java
package com.laixhe.javalog.demo;

/**
 * 数值类型示例：格式化输出、整数溢出安全处理、类型转换。
 * 对应 Rust rustlog/src/number.rs 与 Go golog 中的数值示例。
 *
 * 前置知识：
 * - Java 整数类型：byte(8) short(16) int(32) long(64)，均为有符号
 * - 浮点数：float(32) double(64，默认推荐)
 * - 溢出默认行为：Java 中整数溢出会「静默回绕」，
 *   推荐使用 Math.addExact / Math.multiplyExact 等抛出 ArithmeticException，
 *   或使用 Math.clamp / Math.min / Math.max 实现饱和运算。
 */
public final class NumberDemo {

    private NumberDemo() {
    }

    // ============ 数值转字符串与格式化输出 ============
    public static void numberToString() {
        int i = 666;
        double f1 = 88.888;
        double f2 = 88.0;

        // 基础转字符串
        System.out.println("i=" + i); // 结果：i=666

        // 精度控制（四舍五入）
        System.out.println("f1=" + String.format("%.2f", f1)); // 结果 f1=88.89
        System.out.println("f2=" + String.format("%.2f", f2)); // 结果 f2=88.00

        // ===== 更多格式化方式（新手学习重点）=====

        // 十六进制 / 八进制 / 二进制
        System.out.printf("666 hex=0x%X  octal=0o%o  binary=0b%s%n", i, i, Integer.toBinaryString(i));
        // 结果：666 hex=0x29A  octal=0o1232  binary=0b1010011010

        // 前导零填充 + 宽度控制：%08d 表示「右对齐，总宽度 8，不足补 0」
        System.out.printf("666 with leading zeros: %08d%n", i);
        // 结果：666 with leading zeros: 00000666

        // 对齐：%-10d 左对齐，%10d 右对齐（默认）；居中需要手动补空格
        System.out.printf("left=|%-10d| center=|%6d%4s| right=|%10d|%n", i, i, "", i);
        // 结果：left=|666       | center=|   666    | right=|       666|
        System.out.printf("left=|%-10d| right=|%10d|%n", i, i);

        // 正负号显式显示
        System.out.printf("positive= %+d  negative= %+d%n", 666, -888);
        // 结果：positive= +666  negative= -888

        // 千分位分组
        System.out.printf("grouping: %,d%n", 1_234_567); // 结果：1,234,567
    }

    // ============ 整数溢出安全处理（三种模式对比）============
    // Java 默认整数溢出会「静默回绕」，业务代码应显式选择处理方式
    public static void overflow() {
        int x = Integer.MAX_VALUE;

        // 1) addExact：溢出抛 ArithmeticException，最安全，推荐默认使用
        try {
            int r = Math.addExact(x, 1);
            System.out.println("addExact: MAX+1 = " + r);
        } catch (ArithmeticException e) {
            System.out.println("addExact: MAX+1 = 溢出了（抛出 ArithmeticException）");
        }

        // 2) saturating：饱和运算，溢出时取类型最大值
        //    Java 没有内置 saturating_add，可用 Math.clamp（Java 21+）模拟：
        //    先拓宽到 long 计算避免溢出，再 clamp 回 int 范围
        long wide = (long) x + 1; // 拓宽到 long，不会溢出
        int sat = Math.clamp(wide, Integer.MIN_VALUE, Integer.MAX_VALUE);
        System.out.println("saturating_add: MAX+1 = " + sat + "（饱和，卡在 Integer.MAX_VALUE）"); // 2147483647

        // 3) wrapping：回绕运算（Java 默认行为），适合哈希/CRC 等算法
        int wrap = x + 1;
        System.out.println("wrapping_add: MAX+1 = " + wrap + "（回绕到最小值）"); // -2147483648

        // 4) byte 的显式回绕：127 + 1 = -128
        byte b = Byte.MAX_VALUE;
        byte wrapB = (byte) (b + 1);
        System.out.println("byte wrapping: 127+1 = " + wrapB); // -128

        // ===== 其他常用精确运算方法 =====
        // Math.incrementExact / Math.decrementExact / Math.subtractExact /
        // Math.multiplyExact / Math.negateExact / Math.toIntExact
        try {
            System.out.println("multiplyExact: 100000*100000 = " + Math.multiplyExact(100_000, 100_000));
        } catch (ArithmeticException e) {
            System.out.println("multiplyExact: 100000*100000 = 溢出了（10^10 超出 int 范围）");
        }
    }

    // ============ 数值类型转换 ============
    // 三种转换方式，安全性从高到低：
    // 1) 隐式拓宽（自动）：int -> long -> double，无损失，编译期通过
    // 2) 显式检查转换：Math.toIntExact / 先判断范围，溢出会报错
    // 3) 强制窄化（cast）：可能截断或丢失精度，慎用
    public static void typeConversion() {
        // --- 1. 隐式拓宽（无损失）---
        int small = 10;
        long big = small; // int -> long 自动拓宽，一定安全
        System.out.println("int->long: " + small + " -> " + big); // 10 -> 10

        double d = big; // long -> double 自动拓宽（大数可能丢精度）
        System.out.println("long->double: " + big + " -> " + d);

        // --- 2. 显式检查转换（可能失败，会抛异常）---
        long tooBig = 1000L;
        try {
            int v = Math.toIntExact(tooBig); // long -> int，溢出抛 ArithmeticException
            System.out.println("long->int 成功: " + v);
        } catch (ArithmeticException e) {
            System.out.println("long->int 失败: 1000 -> " + e.getMessage());
        }

        try {
            long overflow = 3_000_000_000L; // 超过 int 范围
            int v = Math.toIntExact(overflow);
            System.out.println("long->int 成功: " + v);
        } catch (ArithmeticException e) {
            System.out.println("long->int 失败: 3000000000 -> 溢出异常：" + e.getClass().getSimpleName());
        }

        // --- 3. as 强制窄化（截断，需慎用）---
        // Java 中 int -> byte 只会保留最低字节
        int a = 0x1234ABCD;
        byte b = (byte) a; // 只保留最低字节 0xCD
        System.out.printf("int as byte 截断: 0x%08X -> 0x%02X%n", a, b & 0xFF); // 0x1234ABCD -> 0xCD

        // 浮点数转整数会向零截断
        double pi = 3.99;
        int truncated = (int) pi;
        System.out.println("double as int 向零截断: " + pi + " -> " + truncated); // 3

        // 如果需要四舍五入，用 Math.round() 再转
        int rounded = (int) Math.round(pi);
        System.out.println("Math.round() 后转: " + pi + " -> " + rounded); // 4

        // 字符串解析（对应 Rust string_parse）
        System.out.println("parseInt('666') = " + Integer.parseInt("666")); // 666
        System.out.println("parseDouble('88.88') = " + Double.parseDouble("88.88")); // 88.88
        try {
            Integer.parseInt("not_a_number");
        } catch (NumberFormatException e) {
            System.out.println("解析失败演示: 'not_a_number' -> NumberFormatException");
        }
    }
}
